### PR TITLE
right shift is now by default sign preserving

### PR DESCRIPTION
**Pull request by krux02**
everything that is needed to make the right bit shit sign preserving by default.

### CODE REVIEW

**Comment by Clyybber** [comment]
This breaks a lot of code, for no benefit.
I think it's better to leave it as it is know (`shr` and `ashr`), as its more explicit and IMO more intuitive.

File: lib/pure/hashes.nim [hunk 1]
```diff
@@ -146,7 +151,8 @@ proc hash*(x: float): Hash {.inline.} =
 
 template bytewiseHashing(result: Hash, x: typed, start, stop: int) =
   for i in start .. stop:
-    result = result !& hash(x[i])
+    let xxx = hash(x[i])
```
Comment (narimiran) [inline lib/pure/hashes.nim]: debugging leftover

**Comment by krux02** [comment]
hmm, ``invalid command line option: '--overflow_checks'`` but only on windows. I wonder how.

File: lib/pure/hashes.nim [hunk 2]
```diff
@@ -60,17 +60,22 @@ proc `!&`*(h: Hash, val: int): Hash {.inline.} =
...
+  result = cast[Hash](res)
 
 proc `!$`*(h: Hash): Hash {.inline.} =
   ## Finishes the computation of the hash value.
   ##
   ## This is only needed if you need to implement a hash proc for a new datatype.
-  result = h +% h shl 3
-  result = result xor (result shr 11)
-  result = result +% result shl 15
+  let h = cast[uint](h) # Hash is practically unsigend.
```
Comment (narimiran) [inline lib/pure/hashes.nim]: typo `unsigend`

File: lib/system.nim [hunk 3]
```diff
@@ -1272,24 +1272,34 @@ else:
...
-    ## .. code-block:: Nim
-    ##   0b0001_0000'i8 shr 2 == 0b0000_0100'i8
-    ##   0b1000_0000'i8 shr 8 == 0b0000_0000'i8
-    ##   0b0000_0001'i8 shr 1 == 0b0000_0000'i8
-  proc `shr`*(x: int8, y: SomeInteger): int8 {.magic: "ShrI", noSideEffect.}
-  proc `shr`*(x: int16, y: SomeInteger): int16 {.magic: "ShrI", noSideEffect.}
-  proc `shr`*(x: int32, y: SomeInteger): int32 {.magic: "ShrI", noSideEffect.}
-  proc `shr`*(x: int64, y: SomeInteger): int64 {.magic: "ShrI", noSideEffect.}
+
+  when defined(arithmeticShiftRight) or true:
```
Comment (krux02) [inline lib/system.nim]: This ``or true`` part is only here for CI testing. It should test what will break if the new right shift is enabled globalley. But the plan is for now to have a transition period where ``shr`` on signed integers triggers a deprecation warning and people will be able to enable arithmetic right shift optionally before it will become the default.

File: compiler/vm.nim [hunk 4]
```diff
@@ -421,12 +421,15 @@ proc opConv(c: PCtx; dest: var TFullReg, src: TFullReg, desttyp, srctyp: PType):
       else:
         let srcDist = (sizeof(src.intVal) - srctyp.size) * 8
         let destDist = (sizeof(dest.intVal) - desttyp.size) * 8
+
+        var value = cast[uint](src.intVal)
```
Comment (Araq) [inline compiler/vm.nim]: I'm pretty sure this should be ``cast[BiggestUInt]`` here.

File: compiler/vm.nim [hunk 5]
```diff
@@ -421,12 +421,15 @@ proc opConv(c: PCtx; dest: var TFullReg, src: TFullReg, desttyp, srctyp: PType):
...
-          dest.intVal = (src.intVal shr srcDist) shl srcDist
-          dest.intVal = (dest.intVal shr destDist) shl destDist
+          value = (value shr srcDist) shl srcDist
+          value = (value shr destDist) shl destDist
         else:
-          dest.intVal = (src.intVal shl srcDist) shr srcDist
-          dest.intVal = (dest.intVal shl destDist) shr destDist
+          value = (value shl srcDist) shr srcDist
+          value = (value shl destDist) shr destDist
+        dest.intVal = cast[int](value)
```
Comment (Araq) [inline compiler/vm.nim]: Likewise, ``cast[BiggestInt]``.

File: compiler/vm.nim [hunk 6]
```diff
@@ -818,7 +821,10 @@ proc rawExecute(c: PCtx, start: int, tos: PStackFrame): TFullReg =
       regs[ra].floatVal = regs[rb].floatVal / regs[rc].floatVal
     of opcShrInt:
       decodeBC(rkInt)
-      regs[ra].intVal = regs[rb].intVal shr regs[rc].intVal
+      let b = cast[uint64](regs[rb].intVal)
+      let c = cast[uint64](regs[rc].intVal)
+      let a = cast[int](b shr c)
```
Comment (Araq) [inline compiler/vm.nim]: ``cast[int64]``!

File: lib/pure/collections/intsets.nim [hunk 7]
```diff
@@ -102,8 +102,9 @@ proc intSetPut(t: var IntSet, key: int): PTrunk =
 proc bitincl(s: var IntSet, key: int) {.inline.} =
   var t = intSetPut(s, `shr`(key, TrunkShift))
   var u = key and TrunkMask
-  t.bits[`shr`(u, IntShift)] = t.bits[`shr`(u, IntShift)] or
-      `shl`(1, u and IntMask)
+  doassert u >= 0
```
Comment (Araq) [inline lib/pure/collections/intsets.nim]: Remove this ``doAssert``, this is performance critical code.

File: lib/pure/collections/intsets.nim [hunk 8]
```diff
@@ -113,11 +114,12 @@ proc exclImpl(s: var IntSet, key: int) =
         dec s.elems
         return
   else:
-    var t = intSetGet(s, `shr`(key, TrunkShift))
+    var t = intSetGet(s, key shr TrunkShift)
     if t != nil:
       var u = key and TrunkMask
-      t.bits[`shr`(u, IntShift)] = t.bits[`shr`(u, IntShift)] and
-          not `shl`(1, u and IntMask)
+      doAssert u >= 0
```
Comment (Araq) [inline lib/pure/collections/intsets.nim]: Same, remove this.

**Comment by Araq** [comment]
Also needs a changelog entry.

**Comment by andreaferretti** [comment]
Apparently, this PR broke [cello](https://github.com/unicredit/cello): I realized only now because there was another issue with nimscript that broke my automated tests.

@krux02  Can you summarize what the changes are and how to get the old behaviour? (I tried to add `-d:oldShiftRight` to my nimbl configuration, but that does not seem to work. Also, this would make the tests pass, but the library would still be broken)

**Comment by andreaferretti** [comment]
@krux02 Could you share any more detail about this change? What should one do to get the old behaviour?

**Comment by narimiran** [comment]
> I tried to add `-d:oldShiftRight` to my nimbl configuration

The correct flag is `-d:nimOldShiftRight`

**Comment by krux02** [comment]
You only use ``shr`` once in the entire library here:
https://github.com/unicredit/cello/blob/1d3edcc4f68443bdac0f372c0ebb644184068d0e/cello.nim#L96

This line needs to be changed to 
``let mask = cast[int](not(0'u) shr (L - i))``
This will work consistent with both old and new Nim.

**Comment by andreaferretti** [comment]
@narimiran @krux02 Unfortunately, even after changing line 96 and adding `-d:nimOldShiftRight` everything still fails. I checked and `shr` is used 11 times acutally. Moreover, I see that `countSetBits` was changed as well, that may be relevant (used 7 times).

In order to fix this, I need to understand *what* this PR has changed. At the end, cello build a few data structures on top of some binary data which is stored as a `seq[int]`. On such ints, I need to perform essentially shifts (`shr`) and popcount (`countSetBits`). Have these two operations changed? Should I use uints to get the old behaviour? A few lines explaining how things were before and how they are now would go a long way. It is very difficult for me to blindly fix this issue

**Comment by Araq** [comment]
> Should I use uints to get the old behaviour?

Yes, this should be ``seq[uint]``.

**Comment by andreaferretti** [comment]
Ok, I should have fixed that. It required a **lot** of going back and forth between uint and int (since uint cannot be used as indices into seqs unfortunately) but that does not seem to alter the performance much. I published version 0.2.0 of cello for Nim 0.20
